google/cloud/pubsub/options.h

```diff
@@ -0,0 +1,170 @@
+ *     will be flushed approximately 5 milliseconds after it is published.
+ */
+struct MaximumHoldTimeOption {
+  using Type = std::chrono::microseconds;
+};
+
+/**
+ * The maximum number of messages in a batch.
+ *
+ * @note Application developers should keep in mind that Cloud Pub/Sub
+ *    sets [limits][pubsub-quota-link] on the size of a batch (1,000 messages)
+ *    The library makes no attempt to validate the value provided in this
+ *    function.
+ *
+ * [pubsub-quota-link]: https://cloud.google.com/pubsub/quotas#resource_limits
+ */
+struct MaximumBatchMessagesOption {
+  using Type = std::size_t;
+};
+
+/**
+ * The maximum size for the messages in a batch.
+ *
+ * @note Application developers should keep in mind that Cloud Pub/Sub
+ *    sets [limits][pubsub-quota-link] on the size of a batch (10MB). The
+ *    library makes no attempt to validate the value provided in this
+ *    function.
+ *
+ * [pubsub-quota-link]: https://cloud.google.com/pubsub/quotas#resource_limits
+ */
+struct MaximumBatchBytesOption {
+  using Type = std::size_t;
+};
+
+/**
+ * The maximum number of pending messages.
+ *
+ * After a publisher flushes a batch of messages the batch is (obviously) not
+ * received immediately by the service. While the batch remains pending it
+ * potentially consumes memory resources in the client (and/or the service).
+ *
+ * Some applications may have constraints on the number of bytes and/or
+ * messages they can tolerate in this pending state, and may prefer to block
+ * or reject messages.
+ */
+struct MaximumPendingMessagesOption {
+  using Type = std::size_t;
+};
+
+/**
+ * The maximum size for pending messages.
+ *
+ * After a publisher flushes a batch of messages the batch is (obviously) not
+ * received immediately by the service. While the batch remains pending it
+ * potentially consumes memory resources in the client (and/or the service).
+ *
+ * Some applications may have constraints on the number of bytes and/or
+ * messages they can tolerate in this pending state, and may prefer to block
+ * or reject messages.
+ */
+struct MaximumPendingBytesOption {
+  using Type = std::size_t;
+};
+
+/**
+ * Publisher message ordering.
+ *
+ * To guarantee messages are received by the service in the same order that
+ * the application gives them to a publisher, the client library needs to wait
+ * until a batch of messages is successfully delivered before sending the next
+ * batch, otherwise batches may arrive out of order as there is no guarantee
+ * the same channel or network path is used for each batch.
+ *
+ * For applications that do not care about message ordering, this can limit
+ * the throughput. Therefore, the behavior is disabled by default.
+ *
+ * @see the documentation for the `Publisher` class for details.
+ */
+struct MessageOrderingOption {
+  using Type = bool;
+};
+
+/// Actions taken by a full publisher.
+enum class FullPublisherAction {
```

Comment:
Hmmm... I was trying to not make these enums public, I cannot recall why. Maybe because I hate `enum`:

* They are the devil to evolve
* They are a code smell for "somebody is using switch when they should be using virtual functions"

But I cannot recall why I did not want them in this case.... Can we hide them in an opaque type?

```cc
class FullPublisherAction {
 public:
   // maybe == and operator!=
  static FullPublisherAction BlockIfFull() { return FullPublisherAction(kBlocks); }
  static FullPublisherAction RejectIfFull() { return FullPublisherAction(kRejects); }
  static FullPublisherAction None() { return FullPublisherAction(kIgnored); }

 private:
   enum Discriminant { kIgnored, kRejects, kBlocks };
   explicit FullPublisherAction(Discriminant d) : discriminant_(d) {}
   friend ... internal::DoStuff(FullPublisherAction a);
};
```

Reply:
I am not convinced that a public enum is a bad idea, but I am not convinced it is a good idea.  Can you provide an argument one way or the other?

Reply:
So quick thoughts are that 

1. user code doesn't change much either way.
```c++
opts.set<FullPublisherActionOption>(FullPublisherAction::kBlockIfFull); // or kBlocks
// vs.
opts.set<FullPublisherActionOption>(FullPublisherAction::BlockIfFull());
```
 
2. Spanner has a similar enum https://github.com/googleapis/google-cloud-cpp/blob/67500487ff3d9c3acb24a5a0228f19a83753443c/google/cloud/spanner/options.h#L117-L125

3. I'm going to guess it's easier to deprecate a method from the class than it is to deprecate a value from the `enum`. Also it seems slightly better to grow a `class` by adding methods than by appending values to an `enum`.

4. Currently, the value is used as an enum. Would it ever grow? For example something that could take in a callback? Like do generic user provided action when Publisher is full, instead of one of the three actions we have listed for you.

5. the class is more complex for us to write. It is ever so slightly more complicated for a user to understand.

Reply:
If there is an action to be taken, other than the ones we list, then the `class` makes sense. Here is a stupid one:

`static FullPublisherAction RejectIfFullWithProbability(double p);`

What might a real action be?

Reply:
Let's go with the public enum.